src/v/archival/ntp_archiver_service.cc

```diff
@@ -854,4 +858,97 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
     co_return result;
 }
 
+void ntp_archiver::rearm_housekeeping_timer() {
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+                          // TODO(vlad): Perhaps use a different
+                          // scheulding group here.
+                          return ss::with_scheduling_group(_upload_sg, [this] {
+                              return apply_retention()
+                                .then([this]() { return garbage_collect(); })
+                                .finally([this] {
+                                    if (_gate.is_closed()) {
+                                        return;
+                                    }
+                                    _housekeeping_timer.rearm(
+                                      _housekeeping_jitter());
+                                });
+                          });
+                      }).handle_exception([this](std::exception_ptr e) {
+        vlog(_rtclog.info, "Error in housekeeping timer callback: {}", e);
+    });
+}
+
+ss::future<> ntp_archiver::apply_retention() {
+    auto log = _partition_manager.log(_ntp);
+    if (!log) {
+        co_return;
+    }
+
+    auto retention_calculator = retention_calculator::factory(
+      manifest(), _partition->get_ntp_config(), *log);
+    if (!retention_calculator) {
+        co_return;
+    }
+
+    auto next_start_offset = retention_calculator->next_start_offset();
+    if (next_start_offset) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
```

Comment:
Do you have a good intuition for how this passes through into remote_partition, when the manifest is updated and stuff gets removed from remote_partition::_segments?

I wonder if it is possible to construct a unit test that intentionally instantiates a reader in flight while then calling into truncate()

Reply:
Good question! Not really. I had a look at the code though. It looks like we lazily update `remote_partition::_segments` from the manifest via `remote_partition::update_segments_incrementally` when making a reader. 

Currently, that only happens if the number of segments in the manifest is larger than the size of `remote_partition::_segments`. The Kafka layer is not aware that we have advanced the start offset (because `remote_partition::_segments` didn't sync with the manifest), so if we get a request below the start offset in the manifest we will try to download the corresponding segment. This will fail as the segment was removed from S3.

Why do we even have `remote_partition::_segments` and not use the manifest as the source of truth? That would make things a lot easier I think.

Reply:
Summarising a discussion from today:
* We can and should use the segments in the manifest as our source of truth. There's some complexity though as remote partitions adjust the offsets to Kafka offsets.
* We should add a unit test for the scenario described in my reply. It will fail for now, but it should start passing when changes to remote partition are made.